### ISSUE TITLE
Password protection of shared dashboards

### ISSUE THREAD

**Issue by EBoisseauSierra** [metabase/metabase#9978]
## Context
Metabase is great. On our instance, we have super-users (admin or not) who have a Metabase account. Those "super-users" create questions and dashboards, and that's all good.

But we also have "viewers only", who shouldn't be able/have to modify dashboards; only looking at dashboards.

## Problem
Sharing dashboards is great for that. However, these dashboards include private/financial data… so we'd better not have them publicly available on the Internet.

[Several options exist](https://discourse.metabase.com/t/best-strategy-to-password-protect-published-dashboards/6159/4) (embedding with Apache protection, LDAP, Apache protection of shared dashboard), but they don't seem fully satisfactory.

## Feature request
_➥ Would it be possible to add the possibility of a **password authentication for dashboards sharing?**_

## Example
You would define the password in the same page where you set up sharing settings, e.g.:
![Screenshot from 2019-05-15 11-27-16_modif_shadow](https://user-images.githubusercontent.com/37387755/57771409-89653380-770a-11e9-9364-048054fdb627.png)

Accessing the URL would then first prompt you for something like this:
![Screenshot from 2019-05-15 12-08-18_shadow](https://user-images.githubusercontent.com/37387755/57771397-7d797180-770a-11e9-8cf2-b42436bebbb7.png)

**Comment by tuaris** [comment]
I've recently run into this issue with the need to provide a partner with access to specific information.  Password protection was a requirement.  As a workaround I had to instead revamp the permissions in our instance, add a user account for the partner, and grant them limited access through group permissions.

I believe that properly configured groups and permissions on the instance is the correct approach.  However it's also understandable that such an approach may be difficult in some circumstances.  Having the option to password protect a shared dashboard would be beneficial.

**Comment by tlrobinson** [comment]
The public URLs are unguessable UUIDs, so it doesn't seem like adding password would add much to the security.

I suppose one argument for passwords is that URLs get saved in browser history.

Regardless of whether we add this, in the meantime you could also build a very simple web application that uses the "Embed this in an application" feature to add password protection, e.x.

```javascript
const express = require("express");
const bodyParser = require("body-parser");
const jwt = require("jsonwebtoken");

const PORT = process.env["PORT"] || 3000;
const METABASE_SITE_URL = process.env["METABASE_SITE_URL"];
const METABASE_SECRET_KEY = process.env["METABASE_SECRET_KEY"];

const PASSWORDS = {
  question: {
    70: "hello"
  },
  dashboard: {
  }
};

const app = express();
app.use(bodyParser.urlencoded({ extended: false }));
app.get("/:type/:id", (req, res) => {
  res.send(`
<form method="post">
    <input name="password" placeholder="Enter the password" />
    <button type="submit">Login</button
</form>`);
});
app.post("/:type/:id", (req, res) => {
  const resourceType = req.params.type;
  const resourceId = parseInt(req.params.id);
  const actualPassworld = req.body.password;
  const expectedPassword = PASSWORDS[resourceType][resourceId];
  if (expectedPassword && expectedPassword === actualPassworld) {
    const payload = {
      resource: { [resourceType]: resourceId },
      params: {},
      exp: Math.round(Date.now() / 1000) + 10 * 60 // 10 minute expiration
    };
    const token = jwt.sign(payload, METABASE_SECRET_KEY);
    res.redirect(`${METABASE_SITE_URL}/embed/${resourceType}/${token}`);
  } else {
    res.send("nope");
  }
});

app.listen(PORT, () => console.log(`Listening on port ${PORT}!`));
```

You could store the passwords in a database somewhere, and have an admin page to add/remove them, etc.

**Comment by EBoisseauSierra** [comment]
@tlrobinson Thanks for your answer. I admit you are unlikely to guess the URL, yet I wouldn't dare telling my client _"Anyone who knows the URL can access your financial data… but you know, it's hard to guess!"_ (-;
In the end, I guess it's a question of "perceived security" (I don't like this concept, though), with a 2FA-ish protection (URL + password).

Anyway, I'll definitively give a try to the code you have sent — I appreciate the time you have granted me!

**Comment by SirTomme** [comment]
We have ProxyPass'ed Metabase through Apache httpd. I tried to protect the Location "/metabase" (and also tried only "/metabase/public") with a basic authorization but it didn't work out well.
Because of the fact that some of Metabase's endpoints return statuses of 401 in some circumstances, Chrome then throws away it's credentials for BasicAuth and repeatedly asks the user to enter username and password for the basic authentication until the whole dashboard is loaded. Firefox although silently re-tries with the credentials the user entered before.  
Any ideas on this? :-)

Finally: I would prefer setting up semi-public dashboards that could then be granted to accounts with no permission to create any question. Something like read-only accounts.

**Comment by EBoisseauSierra** [comment]
@SirTomme I can only say that I had the same issue with 401 responses and so having to enter BasicAuth credentials tons of times.

**Comment by lcaub** [comment]
New to Metabase, but I have the same issue. 🤷 
We need to share data with partners in restricted (filtered only to their data) and protected (if possible with password, or one-time, temporary URL) way.
 
Any update in the last year? 
Alternatives/other tools integrated with metabase?

**Comment by flamber** [comment]
@lcaub That's what Embedding is for: https://www.metabase.com/docs/latest/administration-guide/13-embedding.html
Please use the forum for questions and troubleshooting: https://discourse.metabase.com/

**Comment by lcaub** [comment]
Hey @flamber, thanks for the answer.

At the moment, we don't need to give our partners access to the full set of analytics feature.
We really just want to add a password to a single publicly shared query/dashboard per partner, in read-only mode. 
This partners don't have access to a user account on our platform (unlike customers) so it seems like overkill to develop a whole platform just to use the embedding, which is why I'm digging for other solutions.

I'll wait and see if this feature gets worked on in the future. 👍

**Comment by johan-lejdung** [comment]
I'd love this as a feature as well, embedding is great but just having a simple password would fulfil a need without having to develop a platform just for embedding some statistics.

**Comment by paoliniluis** [comment]
You can easily achieve this by using an NGINX reverse proxy (https://docs.nginx.com/nginx/admin-guide/security-controls/configuring-http-basic-authentication/). You can expose everything that's under /public/* and use basic auth there

**Comment by johan-lejdung** [comment]
True, but that's a lot of maintenance if you need lots of different basic auths. Thanks for the tip though, I might solve it like that for this time. 

Bottom line is still that it would be a convenient and useful feature.

**Comment by easeev** [comment]
Would really appreciate even the simplest implementation of this

Hiding behind a reverse proxy is a viable approach but in case you need different passwords for different shared dashboards easily becomes a management nightmare

**Comment by PekopT** [comment]
This will make Metabase even more enterprise-friendly

**Comment by piyuxhyadav** [comment]
It has been more than 4 years since the issue was raised. Has anyone been able to find a solution?

**Comment by ganeshkrishnan1** [comment]
> You can easily achieve this by using an NGINX reverse proxy (https://docs.nginx.com/nginx/admin-guide/security-controls/configuring-http-basic-authentication/). You can expose everything that's under /public/* and use basic auth there


The problem here is that even on public urls, metabase calls 3 /api calls that bypasses the /public/ url which means /public cannot be exposed out of basic auth easily

**Comment by benjaminwy** [comment]
Agree that this would help with enterprise-friendliness. Our solution is most likely going to be to add viewers as users to our Metabase Cloud account. Ultimately this is more secure anyway, but I don't entirely love it since (1) it incurs per-user costs, and (2) we have to be extra careful when setting permissions so that these view-only clients are only ever allowed to see the specific collections we want them to see, since 99% of our Metabase reports are for internal usage only (and moreover are further restricted based on the department & individual). I just don't love managing external and internal users in the same platform -- feels iffy to me. Still, at the moment this is preferable to setting up a new web app just to embed charts for certain clients. But password-protected sharing URLs would be really nice (like Figma)
